**1. Data Exploration and Preprocessing**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler, LabelEncoder
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.callbacks import EarlyStopping
from itertools import product
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report


df = pd.read_csv('Alphabets_data.csv')
df.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [ ]:
df.shape

(20000, 17)

In [ ]:
df.describe(include='all')

letter          xbox          ybox         width       height  \
count   20000  20000.000000  20000.000000  20000.000000  20000.00000   
unique     26           NaN           NaN           NaN          NaN   
top         U           NaN           NaN           NaN          NaN   
freq      813           NaN           NaN           NaN          NaN   
mean      NaN      4.023550      7.035500      5.121850      5.37245   
std       NaN      1.913212      3.304555      2.014573      2.26139   
min       NaN      0.000000      0.000000      0.000000      0.00000   
25%       NaN      3.000000      5.000000      4.000000      4.00000   
50%       NaN      4.000000      7.000000      5.000000      6.00000   
75%       NaN      5.000000      9.000000      6.000000      7.00000   
max       NaN     15.000000     15.000000     15.000000     15.00000   

               onpix          xbar          ybar         x2bar         y2bar  \
count   20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        3.505850      6.897600      7.500450      4.628600      5.178650   
std         2.190458      2.026035      2.325354      2.699968      2.380823   
min         0.000000      0.000000      0.000000      0.000000      0.000000   
25%         2.000000      6.000000      6.000000      3.000000      4.000000   
50%         3.000000      7.000000      7.000000      4.000000      5.000000   
75%         5.000000      8.000000      9.000000      6.000000      7.000000   
max        15.000000     15.000000     15.000000     15.000000     15.000000   

               xybar       x2ybar        xy2bar         xedge        xedgey  \
count   20000.000000  20000.00000  20000.000000  20000.000000  20000.000000   
unique           NaN          NaN           NaN           NaN           NaN   
top              NaN          NaN           NaN           NaN           NaN   
freq             NaN          NaN           NaN           NaN           NaN   
mean        8.282050      6.45400      7.929000      3.046100      8.338850   
std         2.488475      2.63107      2.080619      2.332541      1.546722   
min         0.000000      0.00000      0.000000      0.000000      0.000000   
25%         7.000000      5.00000      7.000000      1.000000      8.000000   
50%         8.000000      6.00000      8.000000      3.000000      8.000000   
75%        10.000000      8.00000      9.000000      4.000000      9.000000   
max        15.000000     15.00000     15.000000     15.000000     15.000000   

               yedge       yedgex  
count   20000.000000  20000.00000  
unique           NaN          NaN  
top              NaN          NaN  
freq             NaN          NaN  
mean        3.691750      7.80120  
std         2.567073      1.61747  
min         0.000000      0.00000  
25%         2.000000      7.00000  
50%         3.000000      8.00000  
75%         5.000000      9.00000  
max        15.000000     15.00000

In [ ]:
X = df.drop(columns=['letter'])
y = df['letter']
scaler = MinMaxScaler()
X_normalized = pd.DataFrame(scaler.fit_transform(X), columns=X.columns)
X_normalized.describe()

xbox          ybox         width        height         onpix  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       0.268237      0.469033      0.341457      0.358163      0.233723   
std        0.127547      0.220304      0.134305      0.150759      0.146031   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.200000      0.333333      0.266667      0.266667      0.133333   
50%        0.266667      0.466667      0.333333      0.400000      0.200000   
75%        0.333333      0.600000      0.400000      0.466667      0.333333   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

               xbar          ybar         x2bar         y2bar         xybar  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       0.459840      0.500030      0.308573      0.345243      0.552137   
std        0.135069      0.155024      0.179998      0.158722      0.165898   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.400000      0.400000      0.200000      0.266667      0.466667   
50%        0.466667      0.466667      0.266667      0.333333      0.533333   
75%        0.533333      0.600000      0.400000      0.466667      0.666667   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

             x2ybar        xy2bar         xedge        xedgey         yedge  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       0.430267      0.528600      0.203073      0.555923      0.246117   
std        0.175405      0.138708      0.155503      0.103115      0.171138   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.333333      0.466667      0.066667      0.533333      0.133333   
50%        0.400000      0.533333      0.200000      0.533333      0.200000   
75%        0.533333      0.600000      0.266667      0.600000      0.333333   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

             yedgex  
count  20000.000000  
mean       0.520080  
std        0.107831  
min        0.000000  
25%        0.466667  
50%        0.533333  
75%        0.600000  
max        1.000000

In [ ]:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

**2. Model Implementation**

In [ ]:
label_encoder= LabelEncoder()
y_encoded = label_encoder.fit_transform(y)
y_categorical = to_categorical(y_encoded)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_normalized, y_categorical, test_size=0.2, random_state=42,stratify=y_categorical)

In [ ]:
model = Sequential([Dense(32, input_dim = X_train.shape[1], activation='relu'),
Dense(26, activation='softmax')])
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
history = model.fit(X_train, y_train, epochs=20, batch_size=32, validation_split=0.1, verbose=0)
test_loss, test_acc = model.evaluate(X_test, y_test, verbose=0)
print(f'Test accuracy: {test_acc}')
print(f'Test loss: {test_loss}')

Test accuracy: 0.7122499942779541
Test loss: 1.0781539678573608


**3. Hyperparameter Tuning**

In [ ]:
early_stop = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)

def build_and_evaluate(hidden_layers, neurons, activation, learning_rate):
    model = Sequential()
    model.add(Dense(neurons, input_dim=X_train.shape[1], activation=activation))
    for _ in range(hidden_layers - 1):
        model.add(Dense(neurons, activation=activation))
    model.add(Dense(26, activation='softmax'))

    model.compile(optimizer=Adam(learning_rate=learning_rate),
                  loss='categorical_crossentropy', metrics=['accuracy'])

    model.fit(X_train, y_train,
              epochs=5,
              batch_size=32,
              validation_split=0.1,
              callbacks = [early_stop],
              verbose=0)

    loss, acc = model.evaluate(X_test, y_test, verbose=0)
    return acc, loss

hidden_layers_list = [1,2]
neurons_list = [64, 128]
activation_list = ['relu', 'tanh']
learning_rate_list = [0.001, 0.0005]

results = []
total = len(hidden_layers_list) * len(neurons_list) * len(activation_list) * len(learning_rate_list)
for i,(h, n, a, lr) in enumerate(product(hidden_layers_list, neurons_list, activation_list, learning_rate_list),1):
    acc, loss = build_and_evaluate(h, n, a, lr)
    results.append({
        'hidden_layers': h,
        'neurons': n,
        'activation': a,
        'learning_rate': lr,
        'accuracy': acc,
        'loss': loss
    })

sorted_results = sorted(results, key=lambda x: x['accuracy'], reverse=True)
print("\nTop 3 Models:")
for rank, res in enumerate(sorted_results[:3], 1):
    print(f"Rank {rank}:{res}")


Top 3 Models:
Rank 1:{'hidden_layers': 2, 'neurons': 128, 'activation': 'tanh', 'learning_rate': 0.001, 'accuracy': 0.7962499856948853, 'loss': 0.7579541206359863}
Rank 2:{'hidden_layers': 2, 'neurons': 64, 'activation': 'tanh', 'learning_rate': 0.001, 'accuracy': 0.7612500190734863, 'loss': 0.8868126273155212}
Rank 3:{'hidden_layers': 2, 'neurons': 128, 'activation': 'tanh', 'learning_rate': 0.0005, 'accuracy': 0.7484999895095825, 'loss': 0.956615149974823}


**4. Evaluation**

In [ ]:
y_pred = model.predict(X_test)
y_pred_classes = np.argmax(y_pred, axis=1)
y_test_classes = np.argmax(y_test, axis=1)

accuracy = accuracy_score(y_test_classes, y_pred_classes)
precision = precision_score(y_test_classes, y_pred_classes, average='macro')
recall = recall_score(y_test_classes, y_pred_classes, average='macro')
f1 = f1_score(y_test_classes, y_pred_classes, average='macro')
report = classification_report(y_test_classes, y_pred_classes)

print(f'Accuracy: {accuracy}')
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print(f'F1 Score: {f1}')
print('Classification Report:\n', report)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
Accuracy: 0.71225
Precision: 0.7132606999489386
Recall: 0.710496007268245
F1 Score: 0.7059879283804628
Classification Report:
               precision    recall  f1-score   support

           0       0.90      0.85      0.87       158
           1       0.72      0.71      0.71       153
           2       0.67      0.71      0.69       147
           3       0.74      0.70      0.72       161
           4       0.64      0.59      0.61       154
           5       0.71      0.71      0.71       155
           6       0.59      0.45      0.51       155
           7       0.50      0.22      0.30       147
           8       0.89      0.77      0.83       151
           9       0.78      0.79      0.78       149
          10       0.53      0.67      0.59       148
          11       0.90      0.80      0.85       152
          12       0.83      0.92      0.88       158
          13       0.70      0.90      0.79       157
          14       0.

**Default ANN Model Vs Tuned Model**

*  **Improved Accuracy -** Tuned models achieved higher accuracy (up to ~0.79)
compared to the default model's ~0.71, showing better overall classification.

*   **Better Precision & Recall Balance -** The tuned models likely achieved more balanced precision and recall,helping to reduce false positives and false negatives.

* **Reduced Loss -** Loss values improved significantly (~0.75 vs. 1.07 in default), indicating better model confidence and optimization.

* **Optimization of Network Architecture -** Tuning parameters like activation function (tanh), hidden layers and learning rate helped the network learn more effective patterns from data.


